File: src/utils/date.ts
export const JST_OFFSET = 9; // JST is UTC+9, in hours

export const MONTHS = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

export interface Birthday {
  month: number;
  day: number;
}

export function parseMonth(month: string): number {
  return MONTHS.indexOf(month);
}

export function parseBirthday(birthday: string): Birthday | null {
  const [monthStr, dayStr] = birthday.split(" ");

  const month = parseMonth(monthStr);
  const day = Number.parseInt(dayStr);

  if (month === -1 || Number.isNaN(day)) {
    return null;
  }

  return {
    month,
    day,
  };
}

export function dateToJST(date: Date): Date {
  return new Date(date.getTime() + JST_OFFSET * 60 * 60 * 1000);
}

export function currentTimeJST(): Date {
  return dateToJST(new Date());
}

export function currentClosestBreakpointJST(): Date {
  const jstNow = currentTimeJST();

  const year = jstNow.getUTCFullYear();
  const month = jstNow.getUTCMonth();
  const date = jstNow.getUTCDate();
  const hour = jstNow.getUTCHours();

  if (hour < 4) {
    // Before 4:00 JST, use 16:00 JST from the previous day
    return new Date(Date.UTC(year, month, date - 1, 16 - JST_OFFSET, 0, 0));
  }

  if (hour >= 4 && hour < 16) {
    // Between 4:00 and 16:00 JST, use 4:00
    return new Date(Date.UTC(year, month, date, 4 - JST_OFFSET, 0, 0));
  }

  // After 16:00 JST, use 16:00 JST
  return new Date(Date.UTC(year, month, date, 16 - JST_OFFSET, 0, 0));
}

export function resetTimeForDateJST(jstDate: Date): Date {
  const year = jstDate.getUTCFullYear();
  const month = jstDate.getUTCMonth();
  const date = jstDate.getUTCDate();
  const hour = jstDate.getUTCHours();

  if (hour < 4) {
    // Before 4:00 JST, use 4:00 JST from the previous day
    return new Date(Date.UTC(year, month, date - 1, 4 - JST_OFFSET, 0, 0));
  }

  // After 4:00 JST, use 4:00 JST
  return new Date(Date.UTC(year, month, date, 4 - JST_OFFSET, 0, 0));
}

export function currentResetJST() {
  const jstNow = currentTimeJST();
  return resetTimeForDateJST(jstNow);
}

File: src/utils/date.rs
use chrono::{DateTime, Days, NaiveTime, TimeZone, Timelike, Utc};
use chrono_tz::Asia::Tokyo;

pub const MONTHS: [&str; 12] = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
];

/// Birthday with a 0-based month (January = 0), matching JS `Date` months.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Birthday {
    pub month: u32,
    pub day: u32,
}

pub fn parse_month(month: &str) -> Option<u32> {
    MONTHS
        .iter()
        .position(|&m| m == month)
        .map(|i| i as u32)
}

/// Parse a birthday string like `"January 15"`.
pub fn parse_birthday(birthday: &str) -> Option<Birthday> {
    let mut parts = birthday.split_whitespace();
    let month_str = parts.next()?;
    let day_str = parts.next()?;

    let month = parse_month(month_str)?;
    let day = day_str.parse::<u32>().ok()?;

    Some(Birthday { month, day })
}

/// Current instant as `DateTime<Utc>` (wall clock via Asia/Tokyo for JST helpers).
pub fn current_time_jst() -> DateTime<Utc> {
    Utc::now().with_timezone(&Tokyo).with_timezone(&Utc)
}

/// Most recent cafe reset breakpoint (04:00 or 16:00 JST), as UTC.
pub fn current_closest_breakpoint_jst() -> DateTime<Utc> {
    let now = Utc::now().with_timezone(&Tokyo);
    let hour = now.hour();
    let date = now.date_naive();

    let (bp_date, bp_hour) = if hour < 4 {
        (date - Days::new(1), 16u32)
    } else if hour < 16 {
        (date, 4u32)
    } else {
        (date, 16u32)
    };

    let naive = bp_date
        .and_time(NaiveTime::from_hms_opt(bp_hour, 0, 0).expect("valid breakpoint time"));

    Tokyo
        .from_local_datetime(&naive)
        .single()
        .expect("unique Tokyo local time")
        .with_timezone(&Utc)
}

/// Daily reset at 04:00 JST for the JST calendar day of `jst_date`.
/// If the local hour is before 04:00, uses the previous day's 04:00.
pub fn reset_time_for_date_jst(jst_date: DateTime<Utc>) -> DateTime<Utc> {
    let tokyo = jst_date.with_timezone(&Tokyo);
    let hour = tokyo.hour();
    let date = tokyo.date_naive();

    let reset_date = if hour < 4 {
        date - Days::new(1)
    } else {
        date
    };

    let naive = reset_date
        .and_time(NaiveTime::from_hms_opt(4, 0, 0).expect("valid reset time"));

    Tokyo
        .from_local_datetime(&naive)
        .single()
        .expect("unique Tokyo local time")
        .with_timezone(&Utc)
}

/// Daily reset for the current JST day.
pub fn current_reset_jst() -> DateTime<Utc> {
    reset_time_for_date_jst(current_time_jst())
}
